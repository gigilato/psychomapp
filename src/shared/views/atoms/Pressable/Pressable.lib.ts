import { useCallback, useMemo } from 'react'
import { GestureResponderEvent } from 'react-native'
import { useAnimatedStyle, useSharedValue, withTiming, interpolate } from 'react-native-reanimated'

// import * as analytics from '$infra/analytics'

import { IPressableProps } from './Pressable.props'

export const usePressable = ({
  onPressIn,
  onPressOut,
  onPress,
  animation = 'opacity',
  animationOpacity = 0.7,
  animationScale = 0.98,
  animationDuration = 300,
}: Omit<IPressableProps, 'ID'>) => {
  const transition = useSharedValue(0)
  const animationConfig = useMemo(() => ({ duration: animationDuration }), [animationDuration])

  const animatedOnPressIn = useCallback(
    (gestureEvent: GestureResponderEvent) => {
      transition.value = withTiming(1, animationConfig)
      onPressIn?.(gestureEvent)
    },
    [onPressIn, animationConfig, transition]
  )
  const animatedOnPressOut = useCallback(
    (gestureEvent: GestureResponderEvent) => {
      transition.value = withTiming(0, animationConfig)
      onPressOut?.(gestureEvent)
    },
    [onPressOut, animationConfig, transition]
  )

  const handlePress = useCallback(
    (gestureEvent: GestureResponderEvent) => {
      onPress?.(gestureEvent)
      // if (event) analytics.logEvent(event, eventProperties)
    },
    [onPress]
  )

  const animatedStyle = useAnimatedStyle(() => {
    if (animation === 'none') return {}
    const inputRange = [0, 1]

    const isOpacity = animation === 'opacity'
    const value = interpolate(transition.value, inputRange, [
      1,
      isOpacity ? animationOpacity : animationScale,
    ])
    return isOpacity ? { opacity: value } : { transform: [{ scale: value }] }
  }, [animation, animationScale, animationOpacity, transition])

  return {
    onPressIn: animatedOnPressIn,
    onPressOut: animatedOnPressOut,
    onPress: handlePress,
    animatedStyle,
  }
}
