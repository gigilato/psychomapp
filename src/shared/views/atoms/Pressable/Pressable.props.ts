import { PressableProps } from 'react-native'
import { AnimateProps } from 'react-native-reanimated'

import { IShadowProps } from '$views/libs/shadow'

export type PressableAnimation = 'scale' | 'opacity' | 'none'

type NotAnimatedProps = 'onPress' | 'onPressIn' | 'onPressOut' | 'onLongPress'

export interface IPressableProps
  extends Omit<AnimateProps<PressableProps>, NotAnimatedProps>,
    Pick<PressableProps, NotAnimatedProps>,
    IShadowProps {
  ID: string
  animation?: PressableAnimation
  animationOpacity?: number
  animationScale?: number
  animationDuration?: number
  event?: string
  eventProperties?: Record<string, string | undefined>
  // link?: {
  //   type: 'internal' | 'external' | 'authentified'
  //   url: YomoniURL
  //   params?: Record<string, string>
  //   beforeBrowserOpen?: () => boolean
  //   onBrowserClose?: () => void
  // }
}
