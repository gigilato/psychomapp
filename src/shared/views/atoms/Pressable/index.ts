export * from './Pressable'
export * from './Pressable.props'
