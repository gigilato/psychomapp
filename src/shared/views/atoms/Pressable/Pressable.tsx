import { Pressable as PressableDefault } from 'react-native'
import Animated from 'react-native-reanimated'

import { usePressable } from './Pressable.lib'
import { IPressableProps } from './Pressable.props'

const AnimatedPressable = Animated.createAnimatedComponent(PressableDefault)

export const Pressable = ({ style, ID, ...props }: IPressableProps) => {
  const { onPressIn, onPressOut, onPress, animatedStyle } = usePressable(props)
  return (
    <AnimatedPressable
      testID={ID}
      {...props}
      style={[animatedStyle, style]}
      onPressIn={onPressIn}
      onPressOut={onPressOut}
      onPress={onPress}
    />
  )
}
