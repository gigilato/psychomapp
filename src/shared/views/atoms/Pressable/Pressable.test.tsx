import { render } from '@testing-library/react-native'
import { View } from 'react-native'

import { Pressable } from './Pressable'

const Box = () => <View style={{ height: 100, width: 100 }} />

describe('Pressable', () => {
  it('renders correctly', async () => {
    const component = render(
      <Pressable ID="test">
        <Box />
      </Pressable>
    )
    expect(component).toMatchSnapshot()
  })

  it('renders with scale', async () => {
    const component = render(
      <Pressable ID="test" animation="scale">
        <Box />
      </Pressable>
    )
    expect(component).toMatchSnapshot()
  })

  it('renders with styles', async () => {
    const className = 'p-s bg-primary-electric rounded-s'
    const component = render(
      <Pressable ID="test" className={className}>
        <Box />
      </Pressable>
    )
    expect(component).toMatchSnapshot()
  })
})
