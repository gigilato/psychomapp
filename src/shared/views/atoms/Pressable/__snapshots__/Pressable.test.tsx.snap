// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pressable renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      [
        {
          "opacity": undefined,
        },
        undefined,
      ],
    ]
  }
  testID="test"
>
  <View
    style={
      {
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;

exports[`Pressable renders with scale 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  animation="scale"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      [
        {
          "transform": [
            {
              "scale": undefined,
            },
          ],
        },
        undefined,
      ],
    ]
  }
  testID="test"
>
  <View
    style={
      {
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;

exports[`Pressable renders with styles 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "paddingBottom": 12,
        "paddingLeft": 12,
        "paddingRight": 12,
        "paddingTop": 12,
      },
      {
        "backgroundColor": "#FFD700",
      },
      {
        "borderBottomLeftRadius": 6,
        "borderBottomRightRadius": 6,
        "borderTopLeftRadius": 6,
        "borderTopRightRadius": 6,
      },
      [
        {
          "opacity": undefined,
        },
        undefined,
      ],
    ]
  }
  testID="test"
>
  <View
    style={
      {
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;
